### PR TITLE
Lossy Numeric Regex

### PR DESCRIPTION
**Pull request by ivakegg**
original regex for number types

### CODE REVIEW

File: warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java [hunk 1]
```diff
@@ -127,6 +129,7 @@ public static void writeItAll(AccumuloClient client, WhatKindaRange range) throw
             mutation.put(datatype + "\u0000" + corleoneChildUID, "BIRTH_DATE" + "\u0000" + "1930-12-28T00:00:05.000Z", columnVisibility, timeStamp, emptyValue);
             mutation.put(datatype + "\u0000" + corleoneChildUID, "DEATH_DATE" + "\u0000" + "2000-12-28T00:00:05.000Z", columnVisibility, timeStamp, emptyValue);
 
+            mutation.put(datatype + "\u0000" + sopranoUID, "BAIL.0" + "\u0000" + "123401", columnVisibility, timeStamp, emptyValue);
```
Comment (jack-gitdev) [inline warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java]: Nice catch on this edge case!

**Review by jack-gitdev — DISMISSED** [review]

**Review by jack-gitdev — DISMISSED** [review]

**Review by apmoriarty — DISMISSED** [review]

**Review by hlgp — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 2]
```diff
@@ -317,10 +318,15 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
...
                     List<JexlNode> normalizedNodes = Lists.newArrayList();
                     boolean failedNormalization = false;
+                    boolean regexNode = (node instanceof ASTNRNode || node instanceof ASTERNode);
                     // Build up a set of normalized terms using each normalizer
                     for (Type<?> normalizer : dataTypes) {
                         try {
                             if (normalizer instanceof OneToManyNormalizerType && ((OneToManyNormalizerType<?>) normalizer).expandAtQueryTime()) {
+                                if (regexNode) {
+                                    throw new IllegalArgumentException(
+                                                    "OneToManyNormalizers to not handle regex normalization: " + fieldName + " -> " + normalizer.getClass());
```
Comment (hlgp) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: this is no longer accurate.  there are some one-to-many normalizers which can and probably should now.

**Review by hlgp — DISMISSED** [review]

**Review by avgAGB — DISMISSED** [review]

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java [hunk 3]
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
         }
     }
 
+    @Test
+    public void testNumericExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: this can go away

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java [hunk 4]
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
...
+        Assert.assertTrue("Expected equality: " + expectedQueryStr + " vs " + plan,
+                        TreeEqualityVisitor.isEqual(JexlASTHelper.parseJexlQuery(expectedQueryStr), logic.getConfig().getQueryTree()));
+    }
+
+    @Test
+    public void testAnyfieldNumericExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: this can go away

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java [hunk 5]
```diff
@@ -372,6 +383,77 @@ public void testMethodNoExpansion() throws Exception {
...
+        Assert.assertTrue("Expected equality: " + expectedQueryStr + " vs " + plan,
+                        TreeEqualityVisitor.isEqual(JexlASTHelper.parseJexlQuery(expectedQueryStr), logic.getConfig().getQueryTree()));
+    }
+
+    @Test
+    public void testLeadingNumericExpansion() throws Exception {
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        extraParameters.put("hit.list", "true");
+        // extraParameters.put("query.syntax", "LUCENE");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: this can go away

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 6]
```diff
@@ -329,29 +331,55 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
...
                                     normalizedTerms.add(normTerm);
-                                    normalizedNodes.add(JexlNodeFactory.buildUntypedNode(node, fieldName, normTerm));
-                                } else {
-                                    List<JexlNode> normalizedOneToManyNodes = Lists.newArrayList();
-                                    Iterator<String> iter = normTerms.iterator();
-                                    while (iter.hasNext()) {
-                                        normalizedOneToManyNodes.add(JexlNodeFactory.buildUntypedNode(node, fieldName, iter.next()));
+                                    JexlNode normalizedNode = JexlNodeFactory.buildUntypedNode(node, fieldName, normTerm);
+                                    if ((!term.equals(normTerm)) && normalizer.normalizedRegexIsLossy(term)) {
+                                        JexlNode evalOnly = QueryPropertyMarker.create(JexlNodeFactory.buildUntypedNode(node, fieldName, term),
```
Comment (hlgp) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:361]: we need to check if the term is indexOnly before adding EVAL_ONLY node

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:361]: done

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java (hunk 3 above)
Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: done

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java (hunk 4 above)
Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: done

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java (hunk 5 above)
Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExecutableExpansionVisitorTest.java]: done

File: warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java [hunk 7]
```diff
@@ -429,6 +453,10 @@ public static void writeItAll(AccumuloClient client, WhatKindaRange range) throw
             mutation.put("SENTENCE", shard + "\u0000" + datatype, columnVisibility, timeStamp,
                             range == WhatKindaRange.SHARD ? getValueForNuthinAndYourHitsForFree() : getValueForBuilderFor(caponeUID));
             bw.addMutation(mutation);
+            mutation = new Mutation("+dE4.015");
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/util/WiseGuysIngest.java:456]: I apologize for not catching this earlier. This should be `normalizer.normalize()` the numeric value, perhaps with the normalized value in a comment

**Review by apmoriarty — APPROVED** [review]
Look good, one non-blocking comment

**Review by hlgp — APPROVED** [review]
testing looked great
